ensime_shared/ensime.py

```diff
@@ -300,3 +362,13 @@ def fun_en_complete_func(self, client, findstart_and_base, base=None):
             # Invoked by vim
             findstart = findstart_and_base
         return client.complete_func(findstart, base)
+
+    def watchdog_client_server_connection(self, client):
```

Comment:
Ah okay yeah, I hadn't gone that far but now I have while trying to determine if something else was a bug… the latest commit probably gets you even closer (it does some of the same things you started doing on #350 in fact) but I think that `EnsimeProcess` has some wrong behavior that needs to be fixed in order for `Server` to work right when it has a stale reference to a process. I might leave that for you to pick up in the interest of keeping the scope in check here.